syntax = "proto3";

package axiom.generation.v1;

option go_package = "github.com/axiom/api/gen/generation/v1";

// =============================================================================
// GENERATION SERVICE
// Real-time code generation with bidirectional streaming
// =============================================================================

service GenerationService {
  // Bidirectional streaming for interactive generation
  // Client sends intent updates, server streams generation events
  rpc GenerateStream(stream IntentUpdate) returns (stream GenerationEvent);
  
  // Unary generation request (for simple cases)
  rpc Generate(GenerateRequest) returns (GenerateResponse);
  
  // Get generation status
  rpc GetStatus(GetStatusRequest) returns (GetStatusResponse);
  
  // Cancel ongoing generation
  rpc Cancel(CancelRequest) returns (CancelResponse);
}

// =============================================================================
// MESSAGES
// =============================================================================

// Intent update from client (bidirectional streaming)
message IntentUpdate {
  string ivcu_id = 1;
  
  oneof update {
    InitialIntent initial = 2;
    IntentRefinement refinement = 3;
    ContractAddition contract = 4;
    StopGeneration stop = 5;
    SelectCandidate select = 6;
  }
}

message InitialIntent {
  string raw_intent = 1;
  string language = 2;
  string model_id = 3;  // Optional model override
  map<string, string> metadata = 4;
}

message IntentRefinement {
  string refinement_text = 1;
  bool clear_candidates = 2;
}

message ContractAddition {
  string contract_type = 1;  // precondition, postcondition, invariant
  string contract_expression = 2;
  string description = 3;
}

message StopGeneration {
  string reason = 1;
}

message SelectCandidate {
  string candidate_id = 1;
}

// Generation event from server (streaming response)
message GenerationEvent {
  string ivcu_id = 1;
  int64 timestamp = 2;
  
  oneof event {
    GenerationStarted started = 3;
    TokenGenerated token = 4;
    CandidateComplete candidate = 5;
    VerificationProgress verification = 6;
    GenerationComplete complete = 7;
    GenerationError error = 8;
    CostUpdate cost = 9;
  }
}

message GenerationStarted {
  string model_id = 1;
  string model_name = 2;
  string tier = 3;
  float estimated_cost = 4;
}

message TokenGenerated {
  string candidate_id = 1;
  string token = 2;
  int32 token_index = 3;
  bool is_complete = 4;
}

message CandidateComplete {
  string candidate_id = 1;
  string code = 2;
  float confidence = 3;
  string reasoning = 4;
  int32 tokens_used = 5;
}

message VerificationProgress {
  string candidate_id = 1;
  string tier = 2;  // tier_0, tier_1, tier_2, tier_3
  string verifier = 3;
  bool passed = 4;
  float confidence = 5;
  repeated string errors = 6;
  repeated string warnings = 7;
  float execution_time_ms = 8;
}

message GenerationComplete {
  bool success = 1;
  string selected_candidate_id = 2;
  string final_code = 3;
  float overall_confidence = 4;
  int32 total_candidates = 5;
  int32 passing_candidates = 6;
  float total_cost = 7;
  float total_time_ms = 8;
}

message GenerationError {
  string error_code = 1;
  string message = 2;
  bool recoverable = 3;
  string suggested_action = 4;
}

message CostUpdate {
  float current_cost = 1;
  float estimated_remaining = 2;
  string model_id = 3;
  int32 tokens_used = 4;
}

// Unary request/response
message GenerateRequest {
  string raw_intent = 1;
  string language = 2;
  string model_id = 3;
  repeated Contract contracts = 4;
  GenerationOptions options = 5;
}

message Contract {
  string type = 1;
  string expression = 2;
  string description = 3;
}

message GenerationOptions {
  int32 max_candidates = 1;
  int32 timeout_seconds = 2;
  bool run_tier2_verification = 3;
  bool run_tier3_verification = 4;
  float max_cost = 5;
}

message GenerateResponse {
  string ivcu_id = 1;
  bool success = 2;
  string code = 3;
  float confidence = 4;
  VerificationSummary verification = 5;
  CostSummary cost = 6;
}

message VerificationSummary {
  bool tier0_passed = 1;
  bool tier1_passed = 2;
  bool tier2_passed = 3;
  bool tier3_passed = 4;
  float overall_confidence = 5;
  int32 total_errors = 6;
  int32 total_warnings = 7;
}

message CostSummary {
  float total_cost = 1;
  int32 input_tokens = 2;
  int32 output_tokens = 3;
  string model_id = 4;
  int32 generation_attempts = 5;
}

// Status request/response
message GetStatusRequest {
  string ivcu_id = 1;
}

message GetStatusResponse {
  string ivcu_id = 1;
  string status = 2;  // generating, verifying, complete, failed
  int32 candidates_generated = 3;
  int32 candidates_verified = 4;
  float current_cost = 5;
  float elapsed_time_ms = 6;
}

// Cancel request/response
message CancelRequest {
  string ivcu_id = 1;
  string reason = 2;
}

message CancelResponse {
  bool success = 1;
  string message = 2;
}
